synapse/appservice/__init__.py

```diff
@@ -172,12 +172,23 @@ async def _matches_user_in_member_list(
         Returns:
             True if this service would like to know about this room.
         """
-        member_list = await store.get_users_in_room(
+        # We can use `get_local_users_in_room(...)` here because an application
+        # service can only act on behalf of users of the server it's on.
+        #
+        # In the future, we can consider re-using
+        # `store.get_app_service_users_in_room` which is very similar to this
+        # function but has a slightly worse performance than this because we
+        # have an early escape-hatch if we find a single user that the
+        # appservice is interested in. The juice would be worth the squeeze if
+        # `store.get_app_service_users_in_room` was used in more places besides
+        # an experimental MSC. But for now we can avoid doing more work and
+        # barely using it later.
+        local_user_ids = await store.get_local_users_in_room(
```

Comment:
This changes the behavior of `is_interested_in_room` slightly. Previously, we could match remote users when the appservice's user regex didn't include a hostname (all of our test cases and examples in the documentation don't). Was that just a bug?

Reply:
> Was that just a bug?

I don't see how an appservice could control remote users, so I think so...but might be worth asking someone with more info?

Reply:
It's not really a question of controlling the users, but more of whether the application service wants to receive events and updates related to those users, which an application service may well want to do for events coming from non-local users.

So I think this is a change we wouldn't want to go ahead with.

Reply:
Interesting! I hadn't considered this use case.

For my own reference, the spec doesn't mention any restriction either: https://spec.matrix.org/v1.1/application-service-api/#registration

I've updated the comments to clarify local and remote so it's obvious if this is attempted again. We do have tests around this kinda but because they were just mocking `get_users_in_room`, it happily passed when I updated it to `get_local_users_in_room` as well. I've added some actual homeserver tests and verified that they fail if we use `get_local_users_in_room` here ✅

Reply:
Created https://github.com/matrix-org/synapse/pull/14000 to introduce these clarification changes with so we can leave the diff in this PR for reference of what we didn't want to do.

Reply:
The conclusion from https://github.com/matrix-org/synapse/pull/14000#issuecomment-1266552433 is that we don't actually want to support application services being interested in remote users. Thanks for bring it up @erikjohnston and @anoadragon453 for following up in #matrix-spec to clarify it with everyone,

> As @Half-Shot put it, defining remote users is merely a footgun for which an Application Service may assume that it'll receive all events sent by that user, even though it will only receive events in rooms that are shared with a local user.

The path forward is this PR again:

> So by that logic, I'm fine with us defining (and limiting in the code) user namespaces to only cover local users.

Reply:
ftr, such a change would require an MSC I believe - the spec considers it a feature to be able to listen to register remote users in the namespace.

Reply:
That being said, the changes here are valid - simply only returning events targeted at local users covers the same use cases.

Reply:
I'm jumping into this about a week late, but if Synapse doesn't respect `@.*` out of the box then it'd be in violation of the spec - it sounds like this change *causes* such a case.

Reply:
I'm a bit concerned I might not fully understand the impact of the change here - from what I've been reading, it'd be a major spec violation for this to land, though I would be surprised if that were the actual case here. 

Is this literally just treating `@namespace.*` as `@namespace.*:localdomain`, or is there something else going on?

Reply:
@turt2live This PR makes it so `@namespace.*` is only tried against the local users of the server.

Previously, we would try applying the regex to everyone in the room.

Reply:
That would be a major spec violation indeed, which makes this PR not safe to land (with my SCT hat on). We would need *at least* an MSC, which the Synapse team might impose additional restrictions upon (config requirements, etc). 

The full line of investigation on the spec side was to answer the question whether `@namespace.*` actually does match all users regardless of homeserver or if the spec was permissive to restricting to just the local homeserver. In searching the old matrix-doc repo (where the spec & proposals used to live as one, now just proposals), https://github.com/matrix-org/matrix-spec-proposals/issues/1307 came up as part of the r0 effort in 2018/2019, where we attempted to declare the spec out of "beta" and as production-quality. That issue clearly tries to communicate that it is in fact all users regardless of homeserver, and the spec PR reinforces it. What was clearly lacking (in hindsight) is that the spec phrasing itself was not any more helpful as it did not account for the `/user` endpoint or really anything else to do with namespaces. From a spec perspective, this is a regrettable situation but would mean that the spec *intended* to say the namespace can make an appservice interested in remote users, though there's obviously no ability for the server to call `/user` on remote users (it's not like the appservice can create them). Thus, it's specified as such and a clarification PR to solidify that state would be appreciated.

An MSC which changes that would also be good, as clearly it's a great way to cause problems in various ways.

All told, this is a good reminder that the appservice spec is still largely in its pre-2018 state (it didn't get a whole lot of editing for r0) and didn't get the attention it needed. Going back, I probably would have held r0 back in favour of figuring out issues like this and namespaces in general (there's plenty of cases for "wtf, how does this even work"). We're too far along now to go back like that though, making the only option persevering and fixing it :)

(I don't think it'd be a very complicated MSC, which also means it could go through the process relatively quickly - not sure how much this helps with this PR's timelines, though)

Reply:
@turt2live I feel like the [spec issue](https://github.com/matrix-org/matrix-spec-proposals/issues/1307) and PR doesn't clarify as strongly as you're deriving from it.

> you don't just match on the event's sender for the "users" namespace, you also send all events from a room if a user that matches is in that room.

This just talks about the behavior of matching room members and not just event senders, nothing at all clarifying which room members (remote and/or local). I think it could be assumed either way and makes more sense to assume only local from the whole discussion we're having now and our desire to clarify that it should be local only.

Reply:
> though the historical context (which may be lost to time) is still very prevelant here.

Is there something specific to point to here? Or just from your memory?

Reply:
Created a MSC for this: [MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905)


---

And prodding this point in case we can merge earlier,

> That being said, the changes here are valid - simply only returning events targeted at local users covers the same use cases.

I tend to agree with @anoadragon453 here (I think we agree 😬). The use cases of seeing events from remote users (like moderation) is already supported by matching the entire room with the `rooms` namespace regex so this PR should be ok to merge.

We have to rip the band-aid off at some point whether it before or after https://github.com/matrix-org/matrix-spec-proposals/pull/3905 merges. Maybe there is a way to detect people relying on the old behavior? -> https://github.com/matrix-org/synapse/pull/13958#discussion_r992977143

Reply:
[MSC3905](https://github.com/matrix-org/matrix-spec-proposals/pull/3905) is merged and this behavior is ok to change now ✅